src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs

```diff
@@ -0,0 +1,242 @@
+                return (T)(object)Convert.ToSByte((int)lsv);
+            case LazyStringValue lsv when typeof(byte) == typeof(T):
+                return (T)(object)(byte)Convert.ToByte((int)lsv);
+            case long l when typeof(T) == typeof(float):
+                return (T)(object)Convert.ToSingle(l);
+            case long l when typeof(sbyte) == typeof(T):
+                return (T)(object)Convert.ToSByte(l);
+            case long l when typeof(byte) == typeof(T):
+                return (T)(object)Convert.ToByte(l);
+            case float f when typeof(T) == typeof(float):
+                return (T)(object)f;
+            case float f when typeof(sbyte) == typeof(T):
+                return (T)(object)Convert.ToSByte(f);
+            case float f when typeof(byte) == typeof(T):
+                return (T)(object)Convert.ToByte(f);
+            case byte b when typeof(T) == typeof(float):
+                return (T)(object)Convert.ToSingle(b);
+            case byte b when typeof(sbyte) == typeof(T):
+                return (T)(object)Convert.ToSByte(b);
+            case byte b when typeof(byte) == typeof(T):
+                return (T)(object)Convert.ToByte(b);
+            case sbyte sb when typeof(T) == typeof(float):
+                return (T)(object)Convert.ToSingle(sb);
+            case sbyte sb when typeof(sbyte) == typeof(T):
+                return (T)(object)Convert.ToSByte(sb);
+            case sbyte sb when typeof(byte) == typeof(T):
+                return (T)(object)Convert.ToByte(sb);
+            case LazyNumberValue lnv when typeof(T) == typeof(float):
+                return (T)(object)Convert.ToSingle((float)lnv);
+            case LazyNumberValue lnv when typeof(sbyte) == typeof(T):
+                return (T)(object)Convert.ToSByte((sbyte)lnv);
+            case LazyNumberValue lnv when typeof(byte) == typeof(T):
+                return (T)(object)Convert.ToByte((byte)(sbyte)lnv);
+            case JsValue jsn:
+                value = jsn.AsNumber();
+                break;
+        }
+
+        if (value is double d)
+        {
+            if (typeof(T) == typeof(float))
+                return (T)(object)Convert.ToSingle(d);
+
+            if (typeof(sbyte) == typeof(T))
+                return (T)(object)Convert.ToSByte(d);
+
+            if (typeof(byte) == typeof(T))
+                return (T)(object)Convert.ToByte(d);
+        }
+
+        PortableExceptions.Throw<InvalidDataException>($"Type of T is expected to be either {typeof(float)}, {typeof(sbyte)} or {typeof(byte)}. Got {value.GetType().FullName} instead.");
+        return default;
+    }
+}
+
+public static class GenerateEmbeddings
+{
+    // Dimensions (buffer size) from internals of SmartComponents.
+    private const int I1Size = 48;
+    private const int I8Size = 388;
+    private const int F32Size = 1536;
+    
+    [ThreadStatic] // avoid convoys in querying
+    private static ArrayPool<byte> Allocator;
```

Comment:
isn't that a problem at some point to keep it here without any cleanup?

Reply:
Big buffers should be cleaned via GC (cc @redknightlois ); on the other hand, we're releasing the managed buffer after writing to Corax, so it should reuse the buffer again